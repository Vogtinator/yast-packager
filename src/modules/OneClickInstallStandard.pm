#!/usr/bin/perl -w

#
# Module:	OneClickInstallStandard.pm
# Authors:	Lukas Ocilka <locilka@suse.cz>
# Summary:	Module for parsing One Click Install Standard
#		http://en.opensuse.org/Standards/One_Click_Install
#

package OneClickInstallStandard;

use strict;

use XML::Simple;

use YaPI;
use YaST::YCP;

my %config = (
    # evaluate everything as arrays
    ForceArray => 1,
    # remove the the first XML tag
    KeepRoot => 0,
    NoEscape => 1,
    NoIndent => 1,
    ForceContent => 1,
    ContentKey => '-content',
);

our %TYPEINFO;

##
# Converts XML to a list of maps with all repositories described in the XML content.
#
# @param XML content as descibed at http://en.opensuse.org/Standards/One_Click_Install
# @return list <map, <string, any> >
#
# @struct $[
#         "distversion" : "openSUSE Factory",
#         "url" : "full url of the repository (http://.../)",
#         "format" : "yast",
#         "recommended" : true,
#         "description" : "repository description",
#         "localized_description" : $[
#             "en_GB" : "repository description (localized to en_GB)",
#             ...
#         ],
#         "summary" : "repository summary",
#         "localized_summary" : $[
#             "en_GB" : "repository summary (localized to en_GB)",
#             ...
#         ],
#         "name" : "repository name",
#         "localized_name" : $[
#             "en_GB" : "repository name (localized to en_GB)",
#             ...
#         ],
#         "mirrors" : [
#             $[
#                 "url" : "full url of the mirror (http://.../)",
#                 "location" : "?",
#                 "score" : number,
#             ]
#             ...
#         ]
# ]
BEGIN {$TYPEINFO{GetRepositoriesFromXML} = ["function", ["list", ["map", "string", "any"]], "string"];}
sub GetRepositoriesFromXML ($) {
    my $class = shift;

    my $xmlfilecontent = shift || do {
	y2error ("First parameter needs to be a XML content");
	return undef;
    };

    my $xmlparser = new XML::Simple (%config);
    my $data = $xmlparser->XMLin ($xmlfilecontent);
    $xmlparser = undef;

    my $groups = $data->{'group'} || [];
    $data = undef;

    my @repos = ();
    my $repo = {};
    # default language is used when no 'language' string is defined
    my $default_lang = "en_US";

    foreach my $group (@{$groups}) {
	foreach my $one_repo_group (@{$group->{'repositories'}}) {
	    foreach my $one_repo (@{$one_repo_group->{'repository'}}) {
		$repo = {};
		$repo->{'distversion'} = $group->{'distversion'} || "";
		$repo->{'format'} = $one_repo->{'format'} || "";

		# fills up:
		#     * name, description, summary (string)
		#     * localized_name, localized_description, localized_summary (map <string, string>)
		foreach my $key ('name', 'description', 'summary') {
		    foreach my $repo_keys (@{$one_repo->{$key}}) {
			my $lang = $repo_keys->{'lang'} || $default_lang;
			$repo->{'localized_'.$key}->{$lang} = $repo_keys->{'content'} || "";
		    }
		    if (defined $repo->{'localized_'.$key}->{$default_lang}) {
			$repo->{$key} = $repo->{'localized_'.$key}->{$default_lang};
		    }
		}

		my @mirrors = ();
		foreach my $repo_url (@{$one_repo->{'url'}}) {
		    # default url
		    if (! defined $repo_url->{'score'} && ! defined $repo_url->{'location'}) {
			$repo->{'url'} = $repo_url->{'content'} || "";
		    } else {
			push @mirrors, {
			    'url' => $repo_url->{'content'} || "",
			    'score' => $repo_url->{'score'} || "",
			    'location' => $repo_url->{'location'} || "",
			};
		    }
		}
		if (scalar(@mirrors) > 0) {
		    @{$repo->{'mirrors'}} = @mirrors;
		}

		push @repos, $repo;
	    }
	}
    }

    return \@repos;
}

$! = 1;
