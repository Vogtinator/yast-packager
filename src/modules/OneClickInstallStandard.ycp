/**
 * Module:	OneClickInstallStandard.pm
 * Authors:	Lukas Ocilka <locilka@suse.cz>
 * Summary:	Module for parsing One Click Install Standard
 *		http://en.opensuse.org/Standards/One_Click_Install
 */
{
    module "OneClickInstallStandard";

    textdomain "packager";

    import "FileUtils";

    /**
     * Converts XML file to a list of maps with all repositories described in the XML content.
     *
     * @param XML file
     * @return list <map, <string, any> >
     *
     * @struct $[
     *         "distversion" : "openSUSE Factory",
     *         "url" : "full url of the repository (http://.../)",
     *         "format" : "yast",
     *         "recommended" : true,
     *         "description" : "repository description",
     *         "localized_description" : $[
     *             "en_GB" : "repository description (localized to en_GB)",
     *             ...
     *         ],
     *         "summary" : "repository summary",
     *         "localized_summary" : $[
     *             "en_GB" : "repository summary (localized to en_GB)",
     *             ...
     *         ],
     *         "name" : "repository name",
     *         "localized_name" : $[
     *             "en_GB" : "repository name (localized to en_GB)",
     *             ...
     *         ],
     *         "mirrors" : [
     *             $[
     *                 "url" : "full url of the mirror (http://.../)",
     *                 "location" : "?",
     *                 "score" : number,
     *             ]
     *             ...
     *         ]
     * ]
     */
    global list <map <string, any> > GetRepositoriesFromXML (string filename) {
	if (! FileUtils::Exists (filename)) {
	    y2error ("No such file: '%1'", filename);
	    return nil;
	}

	list <map <string, any> > ret = (list <map <string, any> >) SCR::Read (.xmlrepos, filename);

	if (ret == nil) {
	    y2error ("Wrong XML file: %1", filename);
	}

	return ret;
    }
}
