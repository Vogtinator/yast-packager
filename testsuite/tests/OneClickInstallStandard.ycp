{
    include "testsuite.ycp";
    map READ = $[];

    TESTSUITE_INIT ([READ, $[], $[]], nil);
    import "OneClickInstallStandard";
    import "ValueBrowser";

    // see http://en.opensuse.org/Standards/One_Click_Install
    string xml_content = "<metapackage xmlns:os=\"http://opensuse.org/Standards/One_Click_Install\" xmlns=\"http://opensuse.org/Standards/One_Click_Install\">
	<group distversion=\"openSUSE Factory\">
		<repositories>
			<repository recommended=\"true\" format=\"yast\">
				<name>Main Repository</name>
				<summary>This is the main openSUSE catalogue</summary>
				<summary lang=\"en_US\">This is the main openSUSE catalog</summary>
				<description>This is the main openSUSE package repository containing blah blah blah</description>
				<description lang=\"de\">Insert German Here</description>
				<url>http://download.opensuse.org/distribution/SL-OSS-factory/inst-source</url>
				<url score=\"5\" location=\"ie\">http://ftp.heanet.ie/mirrors/ftp.opensuse.org/opensuse/distribution/SL-OSS-factory/inst-source</url>
			</repository>
			<repository recommended=\"false\">
				<name>Some Other Repo</name>
				<summary>Blah Blah</summary>
				<description>blah blah blah</description>
				<url>http://example.com</url>
			</repository>
		</repositories>
		<products>
			<product>
				<name>MyFavouritePackage</name>
				<summary>This is my favourite package</summary>
				<summary lang=\"en_US\">This is my favorite package</summary>
				<description>Blah blah blah</description>
			</product>
			<product type=\"pattern\">
				<name>SomePattern</name>
				<summary>Demo other resolveable</summary>
				<description>Blah blah blah</description>
			</product>
		</products>
	</group>
	<group distversion=\"openSUSE 10.2\">
		<repositories>
			<repository>
				<name>Main Repository</name>
				<summary>This is the main openSUSE catalogue</summary>
				<description>This is the main openSUSE package repository containing blah blah blah</description>
				<url>http://ftp.heanet.ie/mirrors/ftp.opensuse.org/opensuse/distribution/10.2/repo/oss/</url>
			</repository>
		</repositories>
		<products>
			<product>
				<name>MyFavouritePackage</name>
				<summary>This is my favourite package</summary>
				<summary lang=\"en_US\">This is my favorite package</summary>
				<description>Blah blah blah</description>
			</product>
		</products>
	</group>
</metapackage>";
    
    string testany = tostring(OneClickInstallStandard::GetRepositoriesFromXML(xml_content));
    testany = mergestring (splitstring (testany, "["), "[\n");
    testany = mergestring (splitstring (testany, "]"), "\n]");
    DUMP (testany);
}

