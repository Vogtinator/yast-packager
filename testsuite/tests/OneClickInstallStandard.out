Dump	[
Dump	$[
Dump	"description":"This is the main openSUSE package repository containing blah blah blah", "distversion":"openSUSE Factory", "format":"yast", "localized_description":$[
Dump	"de":"Insert German Here", "en_US":"This is the main openSUSE package repository containing blah blah blah"
Dump	], "localized_name":$[
Dump	"en_US":"Main Repository"
Dump	], "localized_summary":$[
Dump	"en_US":"This is the main openSUSE catalog"
Dump	], "mirrors":[
Dump	$[
Dump	"location":"ie", "score":"5", "url":"http://ftp.heanet.ie/mirrors/ftp.opensuse.org/opensuse/distribution/SL-OSS-factory/inst-source"
Dump	]
Dump	], "name":"Main Repository", "summary":"This is the main openSUSE catalog", "url":"http://download.opensuse.org/distribution/SL-OSS-factory/inst-source"
Dump	], $[
Dump	"description":"blah blah blah", "distversion":"openSUSE Factory", "format":"", "localized_description":$[
Dump	"en_US":"blah blah blah"
Dump	], "localized_name":$[
Dump	"en_US":"Some Other Repo"
Dump	], "localized_summary":$[
Dump	"en_US":"Blah Blah"
Dump	], "name":"Some Other Repo", "summary":"Blah Blah", "url":"http://example.com"
Dump	], $[
Dump	"description":"This is the main openSUSE package repository containing blah blah blah", "distversion":"openSUSE 10.2", "format":"", "localized_description":$[
Dump	"en_US":"This is the main openSUSE package repository containing blah blah blah"
Dump	], "localized_name":$[
Dump	"en_US":"Main Repository"
Dump	], "localized_summary":$[
Dump	"en_US":"This is the main openSUSE catalogue"
Dump	], "name":"Main Repository", "summary":"This is the main openSUSE catalogue", "url":"http://ftp.heanet.ie/mirrors/ftp.opensuse.org/opensuse/distribution/10.2/repo/oss/"
Dump	]
Dump	]
